nat-lab/tests/utils/connection_tracker.py

```diff
@@ -52,6 +64,52 @@ class TcpState(Enum):
     LISTEN = "LISTEN"
 
 
+@dataclass()
+class ConnTrackerViolation:
+    """
+    Represents whether it is possible, the result will become valid in the future
+    For example if connection count is currently lower than minimal results, it is
+    possible that program under test will create one, and result will become valid.
+    On the other hand - if we are already over limit - we will never be able to recover.
+
+    This is useful when waiting for valid conntracker state. Conntracker will stop waiting
+    once at least one validator returns non-recoverable invalid result.
+    """
+
+    recoverable: bool
+
+    """
+    This field is designated for debugging efforts, and should specify human-readable details about
+    validation result, especially if validation failed.
+    """
+    reason: str
+
+
+def merge_results(
+    results: list[Optional[ConnTrackerViolation]],
+) -> Optional[ConnTrackerViolation]:
+    violations = list(filter(lambda r: r is not None, results))
+    if len(violations) == 0:
+        return None
+
+    if len(violations) == 1:
+        return violations[0]
+
+    # Need to merge all violations into single validation result
+    reason = (
+        f"There are {len(violations)} violations of conntracker events expectations: "
+    )
+    for violation in violations:
+        if violation is None:
+            continue
+        reason += "\n  "
+        reason += violation.reason
```

Comment:
fixed

Reply:
I think this will not have a new line between the initial message and the list of violations. Should be enough to add `\n` at the end:
```python
   reason = (
        f"There are {len(violations)} violations of conntracker events expectations:\n"
    )
```

Reply:
fixed